### PR TITLE
Add specs for EventConverter

### PR DESCRIPTION
**Pull request by woksin**
Also created Should* extension methods for doing tedious validation

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1140654762299587/1176470140104257)

### CODE REVIEW

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_store_stream_event/a_non_aggregate_event.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+            committed_event = committed_events.a_committed_event(0);
+            event_converter = new EventConverter();
+            stream_position = 3;
+            partition = Guid.NewGuid();
+        };
+
+        Because of = () => result = event_converter.ToStoreStreamEvent(committed_event, stream_position, partition);
+
+        It should_have_the_same_content = () => result.Content.ToString().ShouldEqual(committed_event.Content);
+        It should_represent_the_same_event = () => result.ShouldRepresentTheSameBaseEventAs(committed_event, stream_position, partition);
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_store_stream_event/a_non_aggregate_event.cs:29]: Check `stream_position` and `partition` by itself.

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_store_stream_event/a_non_aggregate_event.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        static CommittedEvent committed_event;
+        static IEventConverter event_converter;
+        static StreamPosition stream_position;
+        static PartitionId partition;
+        static StreamEvent result;
+
+        Establish context = () =>
+        {
+            committed_event = committed_events.a_committed_event(0);
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_store_stream_event/a_non_aggregate_event.cs:20]: Maybe a non-zero value?

File: Specifications/Events.Store.MongoDB/events.cs [hunk 3]
```diff
@@ -1,31 +1,31 @@
...
 {
     public static class events
     {
-        public const string some_event_content = "{\"something\": \"text\"}";
+        public const string some_event_content = "{ \"something\" : \"text\" }";
 
         public static BsonDocument some_event_content_bson_document => BsonDocument.Parse("{\"something\": \"text\"}");
 
-        public static event_builder new_event(uint event_log_sequence_number, uint aggregate_version) => new event_builder(event_log_sequence_number, aggregate_version);
+        public static event_builder new_event(EventLogSequenceNumber event_log_sequence_number, AggregateRootVersion aggregate_version) => new event_builder(event_log_sequence_number, aggregate_version);
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/events.cs:15]: I would call this `new_event_from_aggregate` just so it is a bit more clear in the specs.

File: Specifications/Events.Store.MongoDB/events.cs [hunk 4]
```diff
@@ -1,31 +1,31 @@
...
 
-        public static event_builder new_event_not_from_aggregate(uint event_log_sequence_number) => new event_builder(event_log_sequence_number);
+        public static event_builder new_event_not_from_aggregate(EventLogSequenceNumber event_log_sequence_number) => new event_builder(event_log_sequence_number);
 
-        public static MongoDB.Events.Event an_event(uint event_log_sequence_number, uint aggregate_version) => new event_builder(event_log_sequence_number, aggregate_version).build();
+        public static MongoDB.Events.Event an_event(EventLogSequenceNumber event_log_sequence_number, AggregateRootVersion aggregate_version) => new event_builder(event_log_sequence_number, aggregate_version).build();
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/events.cs:19]: Same here - `an_event_from_aggregate`.

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/event_log_event.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+{
+    public class event_log_event
+    {
+        static Event stored_event;
+        static IEventConverter event_converter;
+        static Runtime.Events.Store.Streams.StreamEvent result;
+
+        Establish context = () =>
+        {
+            stored_event = events.an_event_not_from_aggregate(0);
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/event_log_event.cs:17]: Maybe a non-zero?

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/event_log_event.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            event_converter = new EventConverter();
+        };
+
+        Because of = () => result = event_converter.ToRuntimeStreamEvent(stored_event);
+
+        It should_return_a_committed_event = () => result.Event.ShouldBeOfExactType<CommittedEvent>();
+        It should_have_the_correct_committed_event = () => result.Event.ShouldBeTheSameAs(stored_event);
+        It should_have_the_correct_stream_position = () => result.Position.Value.ShouldEqual(stored_event.EventLogSequenceNumber);
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/event_log_event.cs:25]: This looks off - `should_have_the_correct_STREAM_POSITION` => `Position.Value.ShouldEqual(stored_event.EVENTLOGSEQUENCENUMBER)`

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_event_log_event/an_external_event.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+{
+    public class an_external_event
+    {
+        static CommittedExternalEvent committed_event;
+        static IEventConverter event_converter;
+        static MongoDB.Events.Event result;
+
+        Establish context = () =>
+        {
+            committed_event = committed_events.a_committed_external_event(0, 0);
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_event_log_event/an_external_event.cs:16]: Maybe non-zeros?

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_event_log_event/an_external_event.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+            event_converter = new EventConverter();
+        };
+
+        Because of = () => result = event_converter.ToEventLogEvent(committed_event);
+
+        It should_have_the_same_content = () => result.Content.ToString().ShouldEqual(committed_event.Content);
+        It should_represent_the_same_event = () => result.ShouldRepresentTheSameBaseEventAs(committed_event);
+        It should_not_be_applied_by_aggregate = () => result.Aggregate.WasAppliedByAggregate.ShouldBeFalse();
+        It should_not_come_from_event_horizon = () => result.EventHorizonMetadata.FromEventHorizon.ShouldBeTrue();
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_event_log_event/an_external_event.cs:25]: Something is off here - "should_NOT_come_from_event_horizon" => "FromEventHorizon.ShouldBeTrue()"

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/an_external/event_log_event.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+{
+    public class event_log_event
+    {
+        static Event stored_event;
+        static IEventConverter event_converter;
+        static Runtime.Events.Store.Streams.StreamEvent result;
+
+        Establish context = () =>
+        {
+            stored_event = events.an_external_event(0);
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/an_external/event_log_event.cs:17]: Maybe non-zero?

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/an_external/event_log_event.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+            event_converter = new EventConverter();
+        };
+
+        Because of = () => result = event_converter.ToRuntimeStreamEvent(stored_event);
+
+        It should_return_a_committed_external_event = () => result.Event.ShouldBeOfExactType<CommittedExternalEvent>();
+        It should_have_the_correct_committed_event = () => (result.Event as CommittedExternalEvent).ShouldBeTheSameAs(stored_event);
+        It should_have_the_correct_stream_position = () => result.Position.Value.ShouldEqual(stored_event.EventLogSequenceNumber);
```
Comment (jakhog) [inline Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/an_external/event_log_event.cs:25]: Same here "stream_position" => "EventLogSequenceNumber"

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_store_stream_event/an_external_event.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+            stream_position = 3;
+            partition = Guid.NewGuid();
+        };
+
+        Because of = () => result = event_converter.ToStoreStreamEvent(committed_event, stream_position, partition);
+
+        It should_have_the_same_content = () => result.Content.ToString().ShouldEqual(committed_event.Content);
+        It should_represent_the_same_event = () => result.ShouldRepresentTheSameBaseEventAs(committed_event, stream_position, partition);
+        It should_not_be_applied_by_aggregate = () => result.Aggregate.WasAppliedByAggregate.ShouldBeFalse();
+        It should_not_come_from_event_horizon = () => result.EventHorizonMetadata.FromEventHorizon.ShouldBeTrue();
```
[the remark this answers is not in the document]
Comment (woksin) [reply Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_store_stream_event/an_external_event.cs:31]: Nice catch

[the remark this answers is not in the document]
Comment (woksin) [reply Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_store_stream_event/an_external_event.cs:31]: This should fix it
```suggestion
        It should_come_from_event_horizon = () => result.EventHorizonMetadata.FromEventHorizon.ShouldBeTrue();
```

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/event_log_event.cs (hunk 6 above)
Comment (woksin) [reply Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/event_log_event.cs:25]: I disagree, it makes sense given the context "when converting to runtime stream event a non aggregate event log event". The StreamPosition of that StreamEvent should be equal to the EventLogSequenceNumber of the stored event

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_event_log_event/an_external_event.cs (hunk 8 above)
Comment (woksin) [reply Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_event_log_event/an_external_event.cs:25]: Copy-pasta... This should fix it.
```suggestion
        It should_come_from_event_horizon = () => result.EventHorizonMetadata.FromEventHorizon.ShouldBeTrue();
```

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/an_external/event_log_event.cs (hunk 10 above)
Comment (woksin) [reply Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/an_external/event_log_event.cs:25]: see comment above

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/an_external/event_log_event.cs (hunk 9 above)
Comment (woksin) [reply Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/an_external/event_log_event.cs]: https://github.com/dolittle-runtime/Runtime/pull/407/commits/15a6e6e5d65414f0ec35b0239389219430a20fc4

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_event_log_event/an_external_event.cs (hunk 7 above)
Comment (woksin) [reply Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_event_log_event/an_external_event.cs]: https://github.com/dolittle-runtime/Runtime/pull/407/commits/15a6e6e5d65414f0ec35b0239389219430a20fc4

File: Specifications/Events.Store.MongoDB/ShouldExtensions.cs [hunk 12]
```diff
@@ -43,5 +44,91 @@ public static void ShouldBeTheSameAs(this Dolittle.Versioning.Version version, M
...
+            storedEvent.StreamPosition.ShouldEqual(streamPosition.Value);
+            storedEvent.Partition.ShouldEqual(partition.Value);
+            storedEvent.Metadata.EventLogSequenceNumber.ShouldEqual(committedEvent.EventLogSequenceNumber.Value);
+            storedEvent.ExecutionContext.ShouldBeTheSameAs(committedEvent.ExecutionContext);
+            storedEvent.Metadata.EventSource.ShouldEqual(committedEvent.EventSource.Value);
+            storedEvent.Metadata.Occurred.ShouldEqual(committedEvent.Occurred);
+            storedEvent.Metadata.Public.ShouldEqual(committedEvent.Public);
+            storedEvent.Metadata.TypeId.ShouldEqual(committedEvent.Type.Id.Value);
+            storedEvent.Metadata.TypeGeneration.ShouldEqual(committedEvent.Type.Generation.Value);
+        }
```
[the remark this answers is not in the document]
Comment (woksin) [reply Specifications/Events.Store.MongoDB/ShouldExtensions.cs]: https://github.com/dolittle-runtime/Runtime/pull/407/commits/7176c72ee0d84fa59a90b92c255d03863404aa3e

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/event_log_event.cs (hunk 6 above)
Comment (jakhog) [reply Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/event_log_event.cs:25]: That makes sense, I guess I didn't really understand what ToEventLogEvent was doing

File: Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/unpartitioned_stream_event.cs [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        static StreamId stream;
+
+        static IEventConverter event_converter;
+        static Runtime.Events.Store.Streams.StreamEvent result;
+
+        Establish context = () =>
+        {
+            stream_position = 2;
+            stream = Guid.NewGuid();
+            stored_event = events.a_stream_event_not_from_aggregate(stream_position, Guid.Empty);
```
[the remark this answers is not in the document]
Comment (jakhog) [reply Specifications/Events.Store.MongoDB/Events/for_EventConverter/when_converting_to_runtime_stream_event/a_non_aggregate/unpartitioned_stream_event.cs:23]:
```suggestion
            stored_event = events.a_stream_event_not_from_aggregate(stream_position, Guid.NewGuid());
```

**Comment by jakhog** [comment]
So - based on our internal conversation - my understanding is that we should not use runtime generated random values. So those should be changed back, not to zeroes, but some other value.

**Comment by jakhog** [comment]
Which also means my last suggestion is wrong 👎

**Pull request merged**